share/completions/terraform.fish

```diff
@@ -1,178 +1,205 @@
+
+### providers
+complete -f -c terraform -n __fish_terraform_needs_command -a providers -d "Print tree of modules with their provider requirements"
+complete -f -c terraform -n "__fish_seen_subcommand_from providers" -a "lock mirror schema"
 
 ### refresh
-complete -f -c terraform -n __fish_use_subcommand -a refresh -d 'Update local state file against real resources'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o backup -d 'Path to backup the existing state file'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o input -d 'Ask for input for variables if not directly set'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o lock -d 'Lock the state file when locking is supported'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o lock-timeout -d 'Duration to retry a state lock'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o no-color -d 'If specified, output won\'t contain any color'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o state -d 'Path to a Terraform state file'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o state-out -d 'Path to write state'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o target -d 'Resource to target'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o var -d 'Set a variable in the Terraform configuration'
-complete -f -c terraform -n '__fish_seen_subcommand_from refresh' -o var-file -d 'Set variables from a file'
+complete -f -c terraform -n __fish_terraform_needs_command -a refresh -d "Update local state file against real resources"
+complete -f -c terraform -n "__fish_seen_subcommand_from $apply" -o compact-warnings -d "Show only error summaries"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o backup -d "Path to backup the existing state file"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o input=true -d "Ask for input for variables if not directly set"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o lock=false -d "Don't hold a state lock"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o lock-timeout -d "Duration to retry a state lock"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o no-color -d "If specified, output won't contain any color"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o state -d "Path to a Terraform state file"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o state-out -d "Path to write state"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o target -d "Resource to target"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o var -d "Set a variable in the Terraform configuration"
+complete -f -c terraform -n "__fish_seen_subcommand_from refresh" -o var-file -d "Set variables from a file"
```

Comment:
There is another bug here: `terraform refresh -var-file ` should offer file completions. Same for some others (like `-state`)

You can fix this by replacing `complete -f` with `complete -r`